packages/backend-core/src/configs/configs.ts

```diff
@@ -0,0 +1,224 @@
+import {
+  Config,
+  ConfigType,
+  GoogleConfig,
+  GoogleInnerConfig,
+  OIDCConfig,
+  OIDCInnerConfig,
+  SettingsConfig,
+  SettingsInnerConfig,
+  SMTPConfig,
+  SMTPInnerConfig,
+} from "@budibase/types"
+import { DocumentType, SEPARATOR } from "../constants"
+import { CacheKey, TTL, withCache } from "../cache"
+import * as context from "../context"
+import env from "../environment"
+import environment from "../environment"
+
+// UTILS
+
+/**
+ * Generates a new configuration ID.
+ * @returns {string} The new configuration ID which the config doc can be stored under.
+ */
+export function generateConfigID(type: ConfigType) {
+  return `${DocumentType.CONFIG}${SEPARATOR}${type}`
+}
+
+export async function getConfig<T extends Config>(
+  type: ConfigType
+): Promise<T | undefined> {
+  const db = context.getGlobalDB()
+  try {
+    // await to catch error
+    const config = (await db.get(generateConfigID(type))) as T
+    return config
+  } catch (e: any) {
+    if (e.status === 404) {
+      return
+    }
+    throw e
+  }
+}
+
+export async function save(config: Config) {
+  const db = context.getGlobalDB()
+  return db.put(config)
+}
+
+// SETTINGS
+
+export async function getSettingsConfigDoc(): Promise<SettingsConfig> {
+  let config = await getConfig<SettingsConfig>(ConfigType.SETTINGS)
+
+  if (!config) {
+    config = {
+      _id: generateConfigID(ConfigType.GOOGLE),
+      type: ConfigType.SETTINGS,
+      config: {},
+    }
+  }
+
+  // overridden fields
+  config.config.platformUrl = await getPlatformUrl({
+    tenantAware: true,
+    config: config.config,
+  })
+  config.config.analyticsEnabled = await analyticsEnabled({
+    config: config.config,
+  })
+
+  return config
+}
+
+export async function getSettingsConfig(): Promise<SettingsInnerConfig> {
+  return (await getSettingsConfigDoc()).config
+}
+
+export async function getPlatformUrl(
+  opts: { tenantAware: boolean; config?: SettingsInnerConfig } = {
+    tenantAware: true,
+  }
+) {
+  let platformUrl = env.PLATFORM_URL || "http://localhost:10000"
+
+  if (!env.SELF_HOSTED && env.MULTI_TENANCY && opts.tenantAware) {
+    // cloud and multi tenant - add the tenant to the default platform url
+    const tenantId = context.getTenantId()
+    if (!platformUrl.includes("localhost:")) {
```

Comment:
Could not this be part of the URL set on config, instead of a check in here? It sounds strange to have "dev" code checks in production

Reply:
The platform url has two forms
- system - the base platform url, no tenancy information - just where dns will resolve to. this never changes
- tenant aware - in a multi tenant environment, this url is dynamic based on the tenant being accessed. we enrich this here

We are able to mimic the production url format locally, see: 
https://github.com/Budibase/account-portal/blob/develop/docs/local_development/url_tenancy.md

We want to perform perform the tenant enriching in production and in development environments (when the above url tenancy is active). The only distinguishing factor that we have for this right now is `localhost`. Unless we were to add another flag for `ENABLE_PLATFORM_URL_TENANT_ENRICHMENT` instead, set to false in local development, unless url tenancy is being tested. We could do that, but I'd prefer not to change it right now as this was a copy paste from another location. 

Another thing we could do is default the value of `PLATFORM_URL` in the core env file instead, which is the pattern we are tending towards instead of doing it inside application logic. There are only a few usages of the env field: 
![Screenshot 2023-02-27 at 11 45 10](https://user-images.githubusercontent.com/8755148/221555500-4cafa60c-3518-4eea-a3b0-2dd416781786.png)
The rest of the application uses the value of this helper function. 

If you think it's worth it, could raise a follow up task to address both of these